### ISSUE TITLE
As a maintainer I need to be able to make backend database checking/fixing via SQL

### ISSUE THREAD

**Issue by IanMayo** [serge-web/serge#2396]
For some data-related problems, making changes directly to the SQLite database is the ultimate fall-back.

I need to be able to do things like this:
* [ ] Find all records of with `MessageType` = `MessageCustom`
* [ ] Find records within the above with `message.startDate` containing `2023`
* [ ] Find records within the above with `a234` as  value within `messages.ownAssets[].asset`
* [ ] Produce a list of the count of `message.ownAssets` for all documents.
* [ ] Produce a list of document  references (message.Reference) which have a count of more than 10 in `message.ownAssets` array
* [ ] Find the document with `MessageType` = `InfoType` with the newest (alphabetically last) value in the `_id` field
* [ ] For the above document, find an entity (data.forces.forces.assets[]) with a `uniqid` of  `a234` and replace `location` with `[23, 54]` 
* [ ] Delete all `MessageCustom` documents for which `details.turnNumber`  < 3
* [ ] Delete all `InfoType` documents for which `gameTurn` < 3
* [ ] Delete all `InfoType` documents with `$.phase = 'adjudication'` except for the most recent one

DB Browser version: 3.12.2, using SQLite: 3.35.5

**Comment by rnllv** [comment]
- [x] Find all records of with MessageType = MessageCustom
```
select seq  from 'by-sequence' where json_extract(json, '$.messageType') = 'CustomMessage';
```
![image](https://user-images.githubusercontent.com/77854924/218811765-55acead9-23dc-422e-8d05-e03be0f6810f.png)


*Note: The JSON functions and operators are built into SQLite by default, as of [SQLite version 3.38.0 (2022-02-22)](https://www.sqlite.org/json1.html). Prior to that, they had to be explicitly compiled into the build with additional flags passed to the compiler.*

**Comment by rnllv** [comment]
- [x] Find records within the above with message.startDate containing 2023
```
select json_extract(json, '$.message.startDate')  from 'by-sequence' 
where json_extract(json, '$.messageType') = 'CustomMessage'
and instr(json_extract(json, '$.message.startDate'), '2023') > 0;

```

Attached datasource has no records corresponding to the criteria.

*Note: instr function available as of [SQLite version 3.7.15 (2012-12-12)](https://www.sqlite.org/changes.html#version_3_7_15)*

**Comment by rnllv** [comment]
- [x] Find records within the above with a234 as value within messages.ownAssets[].asset

Use the following query to check for all *asset* in the array:
```
select a.seq, b.value from 'by-sequence' a, json_each(a.json, '$.message.ownAssets') b
where json_extract(a.json, '$.messageType') = 'CustomMessage'
and instr(json_extract(a.json, '$.message.startDate'), '2023') > 0
and json_extract(b.value, '$.asset') = 'a234';
```

Use the following to check only the  first *asset* in the array:
```
select seq, json_extract(json, '$.message.ownAssets')  from 'by-sequence' 
where json_extract(json, '$.messageType') = 'CustomMessage'
and instr(json_extract(json, '$.message.startDate'), '2023') > 0
and json_extract(json, '$.message.ownAssets[0].asset') = 'a234';
```

**Comment by rnllv** [comment]
- [ ] Find records within the above with message.startDate containing 2023

@IanMayo, any difference between items 2 and 4?

**Comment by rnllv** [comment]
- [x] Produce a list of the count of message.ownAssets for all documents.
```
select seq, json_array_length(json, '$.message.ownAssets') from 'by-sequence';
```
![image](https://user-images.githubusercontent.com/77854924/219123750-69c2b43e-d377-4520-9259-5c983ca6bf44.png)

*Note: NULLs are not considered as 0*

**Comment by rnllv** [comment]
- [x] Produce a list of document references (message.Reference) which have a count of more than 10 in message.ownAssets array
```
select seq, json_extract(json, '$.message.Reference') from 'by-sequence'
where json_array_length(json, '$.message.ownAssets') > 10;
```
![image](https://user-images.githubusercontent.com/77854924/219125275-aaa1fd2f-7784-40e0-a31e-310e1565603f.png)

*Note: Output shown for count comparison of more than 1*

**Comment by rnllv** [comment]
- [x] Find the document with MessageType = InfoType with the newest (alphabetically last) value in the doc_id field

```
select seq, doc_id from 'by-sequence' 
where seq in 
    (select seq from 'by-sequence' 
     where json_extract(json, '$.messageType') = 'InfoMessage' 
     order by doc_id desc limit 1);
```
![image](https://user-images.githubusercontent.com/77854924/219362936-15299385-adf2-4453-9aa5-497d4767b135.png)

**Comment by rnllv** [comment]
@IanMayo, this looks to be too generic of an update

>For the above document, find an entity (data.forces.forces.assets[]) with a uniqid of a234 and replace location with [23, 54]

Should we filter further and update only for select roles? (like update assets of only for **f-blue**)

Identify **f-blue** using uniqid
![image](https://user-images.githubusercontent.com/77854924/219372448-986e59a6-1aed-48df-a72a-f87038f53f34.png)

Identify asset element of **f-blue** using internal uniqid
![image](https://user-images.githubusercontent.com/77854924/219372927-8a77dada-3836-4d14-ba34-f619b49e7522.png)

**Comment by IanMayo** [comment]
No, the `uniqid` field for `assets` are globally unique.  There isn't an `a1` per force.   Actually, in the updated database the id is something like `blue.4.33`.

**Comment by rnllv** [comment]
Okay. Then  the query will search all the **assets** arrays to identify the element, and update location.

**Comment by rnllv** [comment]
- [x] For the above document, find an entity (data.forces.forces.assets[]) with a uniqid of a234 and replace location with [23, 54]

```
with
input_values as (
     select
	'a1200' as search_uniq_id,
	'InfoMessage' as message_type,
	'[23, 54]' as updated_location
),
identified_seq_and_path as (
		select a.seq, b.path || '.location' as path
		from 'by-sequence' a, json_tree(a.json, '$.data.forces.forces') b
		where a.seq in
		      (select seq from 'by-sequence'
		      where json_extract(json, '$.messageType') = (select message_type from input_values)
		      order by doc_id desc limit 1)
		and b.fullkey like '$.data.forces.forces[%].assets[%].uniqid'
		and b.value = (select search_uniq_id from input_values)
)
update 'by-sequence'
set json = json_set(json, (select path from identified_seq_and_path), (select json(updated_location) from input_values))
where seq = (select seq from identified_seq_and_path);
```

**Comment by IanMayo** [comment]
> * [x]  For the above document, find an entity (data.forces.forces.assets[]) with a uniqid of a234 and replace location with [23, 54]
> 
> ```
> with
> input_values as (

That all looks a bit scary & fragile @rnllv .  I see the `params` have been pull out at the start - which does seem better UX.  But, are there any simpler ways of approaching the problem?

**Comment by rnllv** [comment]
@IanMayo, *scary*, yep, it is. *fragile*, nope. I've explicitly split up the individual logic components of the query this way so that anyone could edit it later. And I've consulted each and every function I've used here with the official **Sqlite** documentation just to ensure my understanding of these functions is not flawed.

The meat of the query is the **identified_seq_and_path** CTE. This identifies the document and the exact path in the json to be updated.

If you ignore the *where clause* to identify the *seq* (which is the same logic we used for the previous task) in that CTE, the following query represents the main activity happening there.

```
select a.seq, b.path || '.location' as path
from 'by-sequence' a, json_tree(a.json, '$.data.forces.forces') b
where b.fullkey like '$.data.forces.forces[%].assets[%].uniqid'
and b.value = (select search_uniq_id from input_values)
```

The magic happens in the **json_tree** function, which iterates through the entire json structure recursively while extracting its traversal path. Now that we have the path, we filter the path using the expression **b.fullkey like '$.data.forces.forces[%].assets[%].uniqid'**, which is of interest to us, and the *uniqid* to be checked.

Once the seq and path is identified, it's just a matter of updating the particular record using json_set.

**Comment by rnllv** [comment]
I copied the json content of the particular row before update, added it to a git repo, updated the same with the latest value of the document. Below is the diff git generated. 

![image](https://user-images.githubusercontent.com/77854924/219458327-006a5570-a043-4958-b07e-4c1ba7a29ec4.png)

This ensures that no other sections of the json document is affected.

**Comment by rnllv** [comment]
@IanMayo, to further harden the query for the last task please review the following and let me know.

* Consider replacing the filter 
```
b.fullkey   LIKE   '$.data.forces.forces[%].assets[%].uniqid'
```
with
```
REPLACE
(REPLACE
(REPLACE
(REPLACE
(REPLACE
(REPLACE
(REPLACE
(REPLACE
(REPLACE
(REPLACE (b.fullkey, '0', ''),
'1', ''),
'2', ''),
'3', ''),
'4', ''),
'5', ''),
'6', ''),
'7', ''),
'8', ''),
'9', '') = '$.data.forces.forces[].assets[].uniqid'
```
This is because the above *LIKE* filter could match both the following records whereas we need it to match only the first. And since **sqlite3** doesn't support *REGEXP* out of the box this is the second best option we have.

**RECORDS**
$.data.forces.forces[2].assets[2].uniqid
$.data.forces.forces[2 **].plainwrongpath[2** ].assets[2].uniqid




* Consider replacing *JSON_SET* function, used in the **UPDATE** clause, with *JSON_REPLACE* depending on the *overwrite/create* functionality expected for our use-case.  [Here's the official documentation](https://sqlite.org/json1.html#jset). 
This will just be a drop in text replacement.

**Comment by IanMayo** [comment]
Sorry - that is beyond my level of SQL understanding. Could you explain what's happening please?

**Comment by rnllv** [comment]
* For item 1, it boils down to whether the entities in **assets** array contains any other array of entities that might have a property called *uniqid*. If yes, then we'll definitely need to remove the *LIKE* operator. I would suggest let's change it anyways, even at the cost of making the query a little complex, as that would be compatible with any future changes in the data structure.

* For item 2, what's the expectation if the particular entity (with uniqid = 'a123' to be updated for location) does not have a location attribute? Should the query add the location attribute, or do nothing?

**Comment by rnllv** [comment]
* In simpler terms, for item 1, we're trying to locate the path of the uniqid attribute with a SQL *LIKE* expression with % wild character.
In the following expression, the wild character is expected to match any and all index numbers, but it will match text as well.
'$.data.forces.forces[%].assets[%].uniqid'
The right solution would be use a regular expression in place of the wildcard character to restrict the search just to digits, but since **Sqlite3** doesn't support *REGEXP* out of the box, the work around would be to strip out the index numbers and just compare text identifiers and array brackets as in
'$.data.forces.forces[].assets[].uniqid'

**Comment by rnllv** [comment]
- [x] Delete all MessageCustom documents for which details.turnNumber < 3

```
delete from 'by-sequence' where json_extract(json, '$.messageType') = 'CustomMessage' and json_extract(json, '$.details.turnNumber') < 3;
```

**Comment by rnllv** [comment]
- [x] Delete all InfoType documents for which gameTurn < 3

```
delete from 'by-sequence' where json_extract(json, '$.messageType') = 'InfoMessage' and json_extract(json, '$.gameTurn') < 3;
```

**Comment by rnllv** [comment]
- [x] Delete all InfoType documents with $.phase = 'adjudication' except for the most recent one

```
with adjudication_infomessages as (
select seq, rank() over (order by seq desc) seq_rank from 'by-sequence' where json_extract(json, '$.messageType') = 'InfoMessage' and json_extract(json, '$.phase') = 'adjudication')
delete from 'by-sequence' where seq in (select seq from adjudication_infomessages where seq_rank > 1);
```

**Comment by IanMayo** [comment]
Done. Task complete.
